controllers/users_config_map_controller.go

```diff
@@ -0,0 +1,303 @@
+	for _, repoToUpdate := range repositoriesToAdd {
+		shouldGrantPermission := true
+		for _, repoWithPermission := range reposTeamPermissions {
+			if repoToUpdate == repoWithPermission {
+				shouldGrantPermission = false
+				break
+			}
+		}
+
+		if shouldGrantPermission {
+			log.Info("Grant repository permission to the team", "TeamName", teamName, "RepositoryName", repoToUpdate)
+			if err := r.QuayClient.AddReadPermissionsForRepositoryToTeam(r.QuayOrganization, repoToUpdate, teamName); err != nil {
+				log.Error(err, "failed to grant repo permission to the team", "TeamName", teamName, "RepositoryName", repoToUpdate, l.Action, l.ActionAdd)
+				return ctrl.Result{}, err
+			}
+		}
+	}
+
+	// get users in the team
+	usersInTeam := []string{}
+	for _, user := range teamMembers {
+		usersInTeam = append(usersInTeam, user.Name)
+	}
+	log.Info("Users in the team", "TeamName", teamName, "Users", usersInTeam)
+
+	// add users to the team
+	usersToAdd := getUsersToAdd(additionalUsers, usersInTeam)
+	for _, userToAdd := range usersToAdd {
+		log.Info("Add user to the team", "TeamName", teamName, "UserName", userToAdd)
+		if permanentError, err := r.QuayClient.AddUserToTeam(r.QuayOrganization, teamName, userToAdd); err != nil {
+			if !permanentError {
+				log.Error(err, "failed to add user to the team", "TeamName", teamName, "UserName", userToAdd, l.Action, l.ActionAdd)
+				return ctrl.Result{}, err
+			}
+			// if user doesn't exist just log we don't have to fail because of that
+			log.Info(err.Error())
+		}
+	}
+
+	// remove users from the team
+	usersToRemove := getUsersToRemove(additionalUsers, usersInTeam)
+	for _, userToRemove := range usersToRemove {
+		log.Info("Remove user from the team", "TeamName", teamName, "UserName", userToRemove)
+		if err := r.QuayClient.RemoveUserFromTeam(r.QuayOrganization, teamName, userToRemove); err != nil {
+			log.Error(err, "failed to remove user from the team", "TeamName", teamName, "UserName", userToRemove, l.Action, l.ActionDelete)
+			return ctrl.Result{}, err
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func getQuayTeamName(namespace string) string {
+	// quay team allowed chars are : ^[a-z][a-z0-9]+$.
+	// namespace allowed chars are : ^[a-z0-9]([-a-z0-9]*[a-z0-9])?$
+	validNamespaceString := strings.ReplaceAll(namespace, "-", "x")
+
+	if validNamespaceString[0] >= '0' && validNamespaceString[0] <= '9' {
+		validNamespaceString = fmt.Sprintf("x%s", validNamespaceString)
+	}
+
+	return fmt.Sprintf("%sxteam", validNamespaceString)
+}
+
+func getUsersToAdd(additionalUsers, usersInTeam []string) []string {
+	usersToAdd := []string{}
+	for _, userToAdd := range additionalUsers {
+		shouldAdd := true
+		for _, userInTeam := range usersInTeam {
+			if userToAdd == userInTeam {
+				shouldAdd = false
+				break
+			}
+		}
+		if shouldAdd {
+			usersToAdd = append(usersToAdd, userToAdd)
+		}
+	}
+	return usersToAdd
+}
+
+func getUsersToRemove(additionalUsers, usersInTeam []string) []string {
```

Comment:
This function is identical to the one above. If we decide to do the same with image repositories we can have generic lists diff function.

Reply:
created general function

just not sure about name of function